### ISSUE TITLE
open multiple instances of R in the background 

### ISSUE THREAD

**Issue by Moriand** [REditorSupport/vscode-R#735]
Hello, I have installed the R package and its plugins for the interconnect. I am intrigued by the excess consumption of ram.

![alt text](https://i.ibb.co/RNZ1ZYk/Capture.png "Preview")

Is there any way to avoid this?

**Comment by wangyang1749** [comment]
I have the same problem with the languageserve process staring more than one, which causes a lot of memory usage. I don't know under what circumstances  the second or more languageserver process are started.
![image](https://user-images.githubusercontent.com/33304691/127615731-335e9c49-8d69-4425-a413-ad22ad865d43.png)

**Comment by renkun-ken** [comment]
The task process handles the parsing and diagnostics. If you think the number of sessions is too many, please try writing the following in your `~/.Renviron` and see if it works smoothly when you are editing code?

```
R_LANGSVR_POOL_SIZE=0
```

**Comment by wangyang1749** [comment]
Thanks for your reply, I found a problem when I hit Ctrl+S to save, there was a constant languageserve process being created. These processes cannot be shut down automatically. I know there are processes for code diagnostics, but why don't these processes end automatically. This makes a lot of memory.
![image](https://user-images.githubusercontent.com/33304691/127632391-8e0494e6-6e5f-4c7c-a42d-288b21544fdb.png)

**Comment by renkun-ken** [comment]
Would you like to go to output pane and choose "R Language Server" and paste the log of process start and exit here?

**Comment by renkun-ken** [comment]
If the languge client is constantly creating language server processes, there should be many lines like

```
Language Server (12345) started.
```

**Comment by wangyang1749** [comment]
Hi, This is the output of vscode. vscode version   is 1.58.2.
```
R Language Server (7641) started
R Language Server (8230) started
R Language Server (8230) exited with exit code 0
R Language Server (8327) started
R Language Server (10022) started
R Language Server (10038) started
R Language Server (10299) started
R Language Server (10299) exited with exit code 0
R Language Server (10388) started
R Language Server (10701) started
```
```
R version 4.1.0 (2021-05-18)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=zh_CN.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=zh_CN.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=zh_CN.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=zh_CN.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] forcats_0.5.1       stringr_1.4.0       dplyr_1.0.7        
 [4] purrr_0.3.4         readr_1.4.0         tidyr_1.1.3        
 [7] tibble_3.1.2        ggplot2_3.3.5       tidyverse_1.3.1    
[10] BioinfoR_0.0.0.9000

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.7       cellranger_1.1.0 pillar_1.6.1     compiler_4.1.0  
 [5] dbplyr_2.1.1     tools_4.1.0      jsonlite_1.7.2   lubridate_1.7.10
 [9] lifecycle_1.0.0  gtable_0.3.0     pkgconfig_2.0.3  rlang_0.4.11    
[13] reprex_2.0.0     cli_3.0.1        rstudioapi_0.13  DBI_1.1.1       
[17] curl_4.3.2       haven_2.4.1      xml2_1.3.2       withr_2.4.2     
[21] httr_1.4.2       fs_1.5.0         generics_0.1.0   vctrs_0.3.8     
[25] hms_1.1.0        grid_4.1.0       tidyselect_1.1.1 glue_1.4.2      
[29] R6_2.5.0         fansi_0.5.0      readxl_1.3.1     modelr_0.1.8    
[33] magrittr_2.0.1   backports_1.2.1  scales_1.1.1     ellipsis_0.3.2  
[37] rvest_1.0.0      assertthat_0.2.1 colorspace_2.0-2 utf8_1.2.1      
[41] stringi_1.7.3    munsell_0.5.0    broom_0.7.8      crayon_1.4.1  
```

**Comment by wangyang1749** [comment]
Hello, when I use the vscode R plug-in for a while, the system process is as shown in the figure below, and the memory usage is very high.
![image](https://user-images.githubusercontent.com/33304691/127946962-07f21c1a-8131-4a32-a4f0-057f868eb937.png)
When I manually kill these processes, the memory is significantly reduced.
![image](https://user-images.githubusercontent.com/33304691/127947061-b8fac5d7-2cef-4ebe-b213-d3c16d04ed5f.png)

```
R Language Server (4766) started

R Language Server (5102) started
R Language Server (6215) started
[Error - 8:35:08 AM] Request textDocument/documentSymbol failed.
Error: Connection got disposed.
	at Object.dispose (/home/wangyang/.vscode/extensions/reditorsupport.r-lsp-0.1.16/node_modules/vscode-jsonrpc/lib/common/connection.js:1044:27)
	at Object.dispose (/home/wangyang/.vscode/extensions/reditorsupport.r-lsp-0.1.16/node_modules/vscode-languageclient/lib/common/client.js:73:35)
	at /home/wangyang/.vscode/extensions/reditorsupport.r-lsp-0.1.16/node_modules/vscode-languageclient/lib/common/client.js:2320:28
[Error - 8:35:08 AM] Request textDocument/documentLink failed.
Error: Connection got disposed.
	at Object.dispose (/home/wangyang/.vscode/extensions/reditorsupport.r-lsp-0.1.16/node_modules/vscode-jsonrpc/lib/common/connection.js:1044:27)
	at Object.dispose (/home/wangyang/.vscode/extensions/reditorsupport.r-lsp-0.1.16/node_modules/vscode-languageclient/lib/common/client.js:73:35)
	at /home/wangyang/.vscode/extensions/reditorsupport.r-lsp-0.1.16/node_modules/vscode-languageclient/lib/common/client.js:2320:28
[Error - 8:35:08 AM] Request textDocument/colorPresentation failed.
Error: Connection got disposed.
	at Object.dispose (/home/wangyang/.vscode/extensions/reditorsupport.r-lsp-0.1.16/node_modules/vscode-jsonrpc/lib/common/connection.js:1044:27)
	at Object.dispose (/home/wangyang/.vscode/extensions/reditorsupport.r-lsp-0.1.16/node_modules/vscode-languageclient/lib/common/client.js:73:35)
	at /home/wangyang/.vscode/extensions/reditorsupport.r-lsp-0.1.16/node_modules/vscode-languageclient/lib/common/client.js:2320:28
[Error - 8:35:08 AM] Request textDocument/documentSymbol failed.
Error: Connection got disposed.
	at Object.dispose (/home/wangyang/.vscode/extensions/reditorsupport.r-lsp-0.1.16/node_modules/vscode-jsonrpc/lib/common/connection.js:1044:27)
	at Object.dispose (/home/wangyang/.vscode/extensions/reditorsupport.r-lsp-0.1.16/node_modules/vscode-languageclient/lib/common/client.js:73:35)
	at /home/wangyang/.vscode/extensions/reditorsupport.r-lsp-0.1.16/node_modules/vscode-languageclient/lib/common/client.js:2320:28
R Language Server (6215) exited with exit code 0
R Language Server (6384) started
R Language Server (6462) started
R Language Server (6475) started
R Language Server (7126) started
R Language Server (13890) started
R Language Server (14036) started
R Language Server (16111) started
R Language Server (16840) started
R Language Server (17082) started
R Language Server (17082) exited with exit code 0
R Language Server (21154) started
R Language Server (21154) exited with exit code 0
R Language Server (21758) started
R Language Server (22184) started
R Language Server (22197) started
R Language Server (22184) exited with exit code 0
R Language Server (22283) started
R Language Server (22197) exited with exit code 0
R Language Server (22427) started
R Language Server (22429) started
R Language Server (22429) exited from signal SIGKILL
R Language Server (22427) exited from signal SIGKILL
R Language Server (22283) exited from signal SIGKILL
[Error - 10:15:41 AM] Connection to server is erroring. Shutting down server.
[Error - 10:15:41 AM] Connection to server is erroring. Shutting down server.
[Error - 10:15:41 AM] Connection to server is erroring. Shutting down server.
[Error - 10:15:46 AM] Connection to server got closed. Server will not be restarted.
R Language Server (21758) exited from signal SIGKILL
R Language Server (16840) exited from signal SIGKILL
[Error - 10:15:49 AM] Connection to server is erroring. Shutting down server.
[Error - 10:15:49 AM] Connection to server is erroring. Shutting down server.
[Error - 10:15:49 AM] Connection to server is erroring. Shutting down server.
R Language Server (16111) exited from signal SIGKILL
[Error - 10:15:52 AM] Connection to server is erroring. Shutting down server.
[Error - 10:15:52 AM] Connection to server is erroring. Shutting down server.
[Error - 10:15:52 AM] Connection to server is erroring. Shutting down server.
[Error - 10:15:56 AM] Connection to server got closed. Server will not be restarted.
R Language Server (14036) exited from signal SIGKILL
R Language Server (13890) exited from signal SIGKILL
[Error - 10:15:59 AM] Connection to server is erroring. Shutting down server.
[Error - 10:15:59 AM] Connection to server is erroring. Shutting down server.
[Error - 10:15:59 AM] Connection to server is erroring. Shutting down server.
[Error - 10:16:02 AM] Connection to server got closed. Server will not be restarted.
R Language Server (7126) exited from signal SIGKILL
[Error - 10:16:06 AM] Connection to server got closed. Server will not be restarted.
R Language Server (6475) exited from signal SIGKILL
R Language Server (6462) exited from signal SIGKILL
[Error - 10:16:09 AM] Connection to server is erroring. Shutting down server.
[Error - 10:16:09 AM] Connection to server is erroring. Shutting down server.
[Error - 10:16:09 AM] Connection to server is erroring. Shutting down server.
[Error - 10:16:12 AM] Connection to server got closed. Server will not be restarted.
R Language Server (6384) exited from signal SIGKILL
R Language Server (5102) exited from signal SIGKILL
[Error - 10:16:15 AM] Connection to server is erroring. Shutting down server.
[Error - 10:16:15 AM] Connection to server is erroring. Shutting down server.
[Error - 10:16:15 AM] Connection to server is erroring. Shutting down server.
```

**Comment by renkun-ken** [comment]
@wangyang1749 Looks like you are still using `vscode-r-lsp` which is already integrated into vscode-R and should be uninstalled. Please use the latest vscode-R.

Are you working in multi-root workspace? Or there should not be many language server processes started.

**Comment by gowerc** [comment]
Just to add again that I am also seeing this issue (running vscode inside of wsl2 on windows inside a docker container):

![image](https://user-images.githubusercontent.com/8447209/132254460-7cd6dbd6-1b04-42dc-a592-45fab850c560.png)

**Comment by gowerc** [comment]
After some testing I think its relating to windows sleep/hidernation.  I can't say for certain as I haven't found a repeatable way of forcing the issue but it seems correlated with when the laptop goes to sleep. At least I've been running a background job making http get requests from the same container that sometimes errors if the laptop goes to sleep for an extended time which seems to be correlated with the additional spawning process. That being said its not caused by forcing the laptop to sleep but only after leaving the machine alone for an extend period of time. I will run some tests and update accordingly

**Comment by gowerc** [comment]
Ok scrap that, without putting the laptop to sleep one process failed and 3 started:

![image](https://user-images.githubusercontent.com/8447209/132258772-005f800c-c6d1-457a-8785-bb7118c097e1.png)

**Comment by renkun-ken** [comment]
Thanks for providing the info. Looks like the language client is spawning language servers in an unexpected way.

**Comment by ElianHugh** [comment]
There have been a lot of issues with spawn on Windows, such as not passing PATHEXT. So much so that there are [npm packages](https://www.npmjs.com/package/cross-spawn) for cross-platform spawn compatibility. 

Whilst it looks like this is an issue on both Windows and Linux (and Mac?), this might be something to note (I would still prefer spawn over exec).

**Comment by gowerc** [comment]
Hi @renkun-ken  and team, Is there anything I can do / provide to help diagnose & resolve this issue as its becoming a blocker on my current project due to RAM exhaustion :(

**Comment by renkun-ken** [comment]
@gowerc Are you using multi-root workspace, editing untitled files, or files outside your workspace root?

Would you like to "toggle developer tools" and share the logging of "Start language server for ..."?

**Comment by renkun-ken** [comment]
> This might be a separate issue but I am also seeing this in the "output" panel:

It is caused by #468 and fixed via #469. Please install  the latest development version of languageserver.

**Comment by renkun-ken** [comment]
> And the console log as requested: ![image](https://user-images.githubusercontent.com/8447209/137595213-b0acd1e8-5fff-4260-a545-91b12fb5e828.png)

Thanks for posting the image here. It clearly exposes the cause of starting many langaugeserver processes here. When an R file that is outside any workspace folder open in vscode, a standalone languageserver is started for each file. It should be optimized somehow, e.g. all such files share the same languageserver. I'll take a closer look at this.

**Comment by gowerc** [comment]
I guess to be clear I havent done anything explicit to open these files so I'm assuming this is all happening behind the scenes ?

**Comment by renkun-ken** [comment]
> I guess to be clear I havent done anything explicit to open these files so I'm assuming this is all happening behind the scenes ?

Yes. The current language server support for multi-root workspace, untitled files, and external files via multi-server approach is too resource consuming. Let me come up with a better approach.

**Comment by gowerc** [comment]
Currently on my work Mac if I set `R_LANGSVR_POOL_SIZE=1` in my `~/.Renviron` file then upon opening vscode in a directory with an open R file I see:

```
~/Desktop/Work/test_streams
❯ ps -u gowerc | grep R.framework
  503 73175 ??         0:00.24 /Library/Frameworks/R.framework/Resources/bin/exec/R --silent --slave --no-save --no-restore -f /Users/gowerc/.vscode/extensions/ikuyadeu.r-2.3.1/R/help/helpServer.R
  503 73184 ??         0:01.32 /Library/Frameworks/R.framework/Resources/bin/exec/R --quiet --slave -e languageserver::run(port=52670)
  503 73209 ??         0:00.28 /Library/Frameworks/R.framework/Resources/bin/exec/R --no-readline --slave --no-save --no-restore
  503 73211 ??         0:00.49 /Library/Frameworks/R.framework/Resources/bin/exec/R --no-readline --slave --no-save --no-restore
```

So I guess its behaving as expected :) 

Will test on my windows machine this evening.

What are the consequences / negatives then of using `R_LANGSVR_POOL_SIZE=1` ?

**Comment by albertosantini** [comment]
My context:

- Windows 10 - Version 21H2
- VSCode 1.63.2
- R extension 2.3.4
- R 4.1.2
- languageserver 0.3.12
- Are you using multi-root workspace - No
- editing untitled files - No
- files outside your workspace root - No
- R_LANGSVR_POOL_SIZE=0 - in .Renviron
- "terminal.integrated.enablePersistentSessions": false - in VSCode settings
- "terminal.integrated.persistentSessionReviveProcess": "never" - in VSCode settings

Use cases:

- Open VSCode, open a folder, open a R file, open a Terminal (NOT R Terminal)
  - In Process Explorer (under Visual Studio Code process): there are a few "R for Windows terminal front-end" and "R for Windows front-end" processes (I think thatnumber depends about the configuration above and cpus).
  - Closing VSCode, those processes are gone
  - **OK**

- Open VSCode, open a folder, open a R file, open a Terminal (NOT R Terminal)
  - Execute "R: Create R Terminal"
  - Closing VSCode, those processes are gone
  - **OK**

Then I tried to open other R files and a few different combinations of user actions, trying to get orphan processes when I close VSCode. I cannot repro it, even if, in the past, I saw sometimes a few ones.

Last thing. Generally speaking, one of the processes related to R consumes about 2GB of RAM in my context, just only with the file editor opened or the R Terminal just created.

**Comment by Yunuuuu** [comment]
Likewise, it seems `vscode-r` never terminate the `R front-end`, so when we restart vscode multiple times, the `R for Windows front-end` accumulates gradually.

My running environment:
- OS: `Windows 10 20H2-19042.1415`
- vscode
`Version: 1.63.2 (user setup)`
Commit: 899d46d82c4c95423fb7e10e68eba52050e30ba3
Date: 2021-12-15T09:40:02.816Z
Electron: 13.5.2
Chromium: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.19042
- radian version: `0.5.13`
- R version: `4.1.1`
- languageserver: `0.3.12`
- Are you using multi-root workspace? - `No`
- files outside your workspace root - `No`
- "terminal.integrated.enablePersistentSessions": `false` - in VSCode settings
- "terminal.integrated.persistentSessionReviveProcess": `"never"` - in VSCode settings

1. when I seted `R_LANGSVR_POOL_SIZE = 0` in my `~/.Renviron` and opened a vscode instance (never opened any files), then quit vscode, these process remained runing:
![image](https://user-images.githubusercontent.com/19575010/148586540-6366174a-f9d1-4cf7-90b4-114d52793e3a.png)

2. when I seted `R_LANGSVR_POOL_SIZE = 1` in my `~/.Renviron` and opened a vscode instance (never opened any files), then quit vscode, these process remained runing:
![image](https://user-images.githubusercontent.com/19575010/148587082-3300b9b5-a8e6-4a51-8f47-2ae2bf1b188d.png)

3. when I reopen vscode several times, the process `R for Windows front-end` in `vscode` will accumulate:
![image](https://user-images.githubusercontent.com/19575010/148587516-ef715c01-b138-43a9-837a-aa42a1300685.png)

This shouldn't be caused by `radian`, I have run `radian` alone with `R` in windows `cmd.exe`, which didn't present above results.

**Comment by albertosantini** [comment]
Resolved with https://github.com/REditorSupport/vscode-R/pull/941

**Comment by gowerc** [comment]
Thanks @renkun-ken and @albertosantini for tracking down and fixing this bug :) will make a big difference for windows users !
